### PR TITLE
Feature/ctb 7/add project common error response model to represent application errors in http response

### PR DESCRIPTION
**Pull request by YauhenIlyin**

### CODE REVIEW

File: core/src/main/java/by/ilyin/core/service/CustomUserService.java [hunk 1]
```diff
@@ -31,21 +32,31 @@ public class CustomUserService {
...
         return new CreateUserResponseDTO(realUser.getId());
     }
 
     @Transactional
     public void updateUser(Long id, UpdateUserRequestDTO updateUserRequestDTO) {
-        Optional<CustomUser> optionalCustomUser;
-        optionalCustomUser = customUserRepository.findById(id);
-        CustomUser customUser = optionalCustomUser.orElseThrow(); //todo catch this exception
+        Optional<CustomUser> currentOptionalUser = customUserRepository.findById(id);
+        currentOptionalUser.orElseThrow(() -> new ResourceNotFoundException("Update failed. User with id " + id + " was not found."));
```
Comment (DzmitryKazakevich) [inline core/src/main/java/by/ilyin/core/service/CustomUserService.java]: @YauhenIlyin Why logic with Optional is written in 2 lines?

File: core/src/main/java/by/ilyin/core/service/CustomUserService.java [hunk 2]
```diff
@@ -31,21 +32,31 @@ public class CustomUserService {
...
+        Optional<CustomUser> optionalUserByLogin = customUserRepository.findByLogin(updateUserRequestDTO.getLogin());
+        boolean isCorrect = userValidator.isCorrectLoginWithId(currentOptionalUser, optionalUserByLogin);
```
Comment (DzmitryKazakevich) [inline core/src/main/java/by/ilyin/core/service/CustomUserService.java]: @YauhenIlyin isCorrectLogin logic should be implement using Validator interface. We need to validate request before service specific logic execution.

File: core/src/main/java/by/ilyin/core/service/CustomUserService.java [hunk 3]
```diff
@@ -117,7 +128,7 @@ public Page<CustomUser> getUsers(
 
     public CustomUser getUser(Long id) {
         Optional<CustomUser> optionalUser = customUserRepository.findById(id);
-        return optionalUser.orElseThrow(); //todo catch this exception
+        return optionalUser.orElseThrow(() -> new ResourceNotFoundException("User with id " + id + " was not found."));
```
Comment (DzmitryKazakevich) [inline core/src/main/java/by/ilyin/core/service/CustomUserService.java]: @YauhenIlyin it should be done in 1 line

File: core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    private ResponseEntity<ErrorResponse> buildSimpleErrorResponse(Exception e, HttpStatus httpStatus) {
+        return buildSimpleErrorResponse(e, httpStatus, e.getMessage());
+    }
+
+    private ResponseEntity<ErrorResponse> buildSimpleErrorResponse(Exception e, HttpStatus httpStatus, String message) {
+        return new ResponseEntity<>(
+                new ErrorResponse(
+                        e.getStackTrace(),
```
Comment (DzmitryKazakevich) [inline core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java]: @YauhenIlyin users on product should never see StackTrace!

File: core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                        httpStatus,
```
Comment (DzmitryKazakevich) [inline core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java:106]: @YauhenIlyin For what reason you have http status in error response DTO?

File: web/src/main/java/by/ilyin/web/util/validator/UserValidator.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.validation.BindingResult;
+import org.springframework.validation.ObjectError;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+public class UserValidator {
+
+    public void userValidationProcess(BindingResult bindingResult) {
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/util/validator/UserValidator.java]: @YauhenIlyin What is the purpose of this method?
Did you investigate how BindingResult works?
Why Validator interface is missing?

File: web/src/main/java/by/ilyin/web/util/error/RealErrorResponse.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package by.ilyin.web.util.error;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+public class RealErrorResponse {
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/util/error/RealErrorResponse.java]: @YauhenIlyin What is shitty class name?
What is Real and what is unreal?

File: web/src/main/java/by/ilyin/web/util/error/CustomControllerAdvice.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    //todo add javadoc 404
+
+    @ExceptionHandler(ResourceNotFoundException.class)
+    public ResponseEntity<RealErrorResponse> handleResourceNotFoundExceptions(Exception e) {
+        return buildSimpleErrorResponse(e, HttpStatus.NOT_FOUND);
+    }
+
+    //todo add javadoc 405
+
+    @ExceptionHandler(HttpRequestMethodNotSupportedException.class)
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/util/error/CustomControllerAdvice.java]: @YauhenIlyin Why do you handle it?

File: web/src/main/java/by/ilyin/web/util/error/CustomControllerAdvice.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    //todo add javadoc 409
+
+    @ExceptionHandler(ResourceAlreadyExists.class)
+    public ResponseEntity<RealErrorResponse> handleResourceAlreadyExists(Exception e) {
+        return buildSimpleErrorResponse(e, HttpStatus.CONFLICT);
+    }
+
+    //todo add javadoc 415
+
+    @ExceptionHandler(HttpMediaTypeNotSupportedException.class)
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/util/error/CustomControllerAdvice.java]: @YauhenIlyin  Why do you handle it?

File: web/src/main/java/by/ilyin/web/util/error/CustomControllerAdvice.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    //todo add javadoc 405
+
+    @ExceptionHandler(HttpRequestMethodNotSupportedException.class)
+    public ResponseEntity<RealErrorResponse> handleHttpRequestMethodNotSupportedExceptions(Exception e) {
+        return buildSimpleErrorResponse(e, HttpStatus.METHOD_NOT_ALLOWED);
+    }
+
+    //todo add javadoc 406
+
+    @ExceptionHandler(HttpMediaTypeNotAcceptableException.class)
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/util/error/CustomControllerAdvice.java]: @YauhenIlyin  Why do you handle it?

File: web/src/main/java/by/ilyin/web/feign/decoder/CustomErrorDecoder.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+@Component
+public class CustomErrorDecoder implements ErrorDecoder {
+
+    private final ErrorDecoder errorDecoder = new Default();
+
+    public Exception decode(String methodKey, Response response) {
+        FeignErrorResponse feignErrorResponse;
+        try {
+            InputStream bodyIs = response.body().asInputStream();
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/feign/decoder/CustomErrorDecoder.java]: @YauhenIlyin bodyIs?????

File: web/src/main/java/by/ilyin/web/feign/decoder/CustomErrorDecoder.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            ObjectMapper mapper = new ObjectMapper();
+            feignErrorResponse = mapper.readValue(bodyIs, FeignErrorResponse.class);
+        } catch (IOException e) {
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/feign/decoder/CustomErrorDecoder.java:23]: @YauhenIlyin why default logic is applied only to IOException?

File: web/src/main/java/by/ilyin/web/feign/decoder/CustomErrorDecoder.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+public class CustomErrorDecoder implements ErrorDecoder {
+
+    private final ErrorDecoder errorDecoder = new Default();
+
+    public Exception decode(String methodKey, Response response) {
+        FeignErrorResponse feignErrorResponse;
+        try {
+            InputStream bodyIs = response.body().asInputStream();
+            ObjectMapper mapper = new ObjectMapper();
+            feignErrorResponse = mapper.readValue(bodyIs, FeignErrorResponse.class);
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/feign/decoder/CustomErrorDecoder.java]: @YauhenIlyin 
You can use following code as do what you did in a few lines of code:
FeignException feignException = FeignException.errorStatus(methodKey, response);

Why did you write overhead

File: web/src/main/java/by/ilyin/web/exception/http/ErrorResponseWrapperException.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+import by.ilyin.web.util.error.FeignErrorResponse;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class ErrorResponseWrapperException extends RuntimeException {
+
+    private FeignErrorResponse feignErrorResponse;
```
Comment (DzmitryKazakevich) [inline web/src/main/java/by/ilyin/web/exception/http/ErrorResponseWrapperException.java]: @YauhenIlyin exception contains DTO???

File: core/src/main/java/by/ilyin/core/util/validation/UserValidator.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        return customUserRepository.findByLogin(login).isPresent();
+    }
+
+    public boolean isCorrectLoginWithId(Long id, String login) {
+        return isCorrectLoginWithId(
+                customUserRepository.findById(id),
+                customUserRepository.findByLogin(login));
+    }
+
+    public boolean isCorrectLoginWithId(Optional<CustomUser> optionalUserById, Optional<CustomUser> optionalUserByLogin) {
```
Comment (DzmitryKazakevich) [inline core/src/main/java/by/ilyin/core/util/validation/UserValidator.java]: @YauhenIlyin it can be done much simpler

**Review by DzmitryKazakevich — CHANGES_REQUESTED** [review]

File: core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java (hunk 5 above)
Comment (Egor-Duvalov) [reply core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java:106]: @DzmitryKazakevich it's specified in the task 
image

File: core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java (hunk 4 above)
Comment (YauhenIlyin) [reply core/src/main/java/by/ilyin/core/util/error/CustomControllerAdvice.java]: users will not see the stack trace

**Pull request merged**